### PR TITLE
Add Azure Developer Credential

### PR DESCRIPTION
**Pull request by hemarina**
- add azure developer credential and implementation
- add unit test
- change readme
- update changelog

Note: We would update readme once we have our documentation. This PR will not include it as it's not a blocker.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredentialOptions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+        /// <summary>
+        /// Specifies tenants in addition to the specified <see cref="TenantId"/> for which the credential may acquire tokens.
+        /// Add the wildcard value "*" to allow the credential to acquire tokens for any tenant the logged in account can access.
+        /// If no value is specified for <see cref="TenantId"/> this option will have no effect, and the credential will acquire tokens for any requested tenant.
+        /// </summary>
+        public IList<string> AdditionallyAllowedTenants => AdditionallyAllowedTenantsCore;
+
+        /// <summary>
+        /// The Cli process timeout.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredentialOptions.cs:27]:
```suggestion
        /// The CLI process timeout.
```

**Review by scottaddie — COMMENTED** [review]
- In the README's **Credential classes** -> **Authenticate via development tools** section, add a table row for the new credential type.
- If the new credential type is included in the `DefaultAzureCredential` auth flow by default, then do the following:
	- Add a new list item to the README's **DefaultAzureCredential** section.
	- Update the mermaid.js diagram at https://raw.githubusercontent.com/Azure/azure-sdk-for-net/main/sdk/identity/Azure.Identity/images/mermaidjs/DefaultAzureCredentialAuthFlow.md. See the SVG file generation instructions in the comment at the top of the Markdown file.

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        //TODO 1
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
+        {
+            return GetTokenImplAsync(false, requestContext, cancellationToken).EnsureCompleted();
+        }
+
+        /// <summary>
+        /// Obtains a access token from Azure Developer CLI service, using the access token to authenticate. This method id called by Azure SDK clients.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:87]:
```suggestion
        /// Obtains an access token from Azure Developer CLI service, using the access token to authenticate. This method is called by Azure SDK clients.
```

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredentialOptions.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        /// <summary>
+        /// The ID of the tenant to which the credential will authenticate by default. If not specified, the credential will authenticate to any requested tenant, and will default to the tenant provided to the 'azd login' command.
+        /// </summary>
+        public string TenantId { get; set; }
+
+        /// <summary>
+        /// Specifies tenants in addition to the specified <see cref="TenantId"/> for which the credential may acquire tokens.
+        /// Add the wildcard value "*" to allow the credential to acquire tokens for any tenant the logged in account can access.
+        /// If no value is specified for <see cref="TenantId"/> this option will have no effect, and the credential will acquire tokens for any requested tenant.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredentialOptions.cs:22]:
```suggestion
        /// If no value is specified for <see cref="TenantId"/>, this option will have no effect, and the credential will acquire tokens for any requested tenant.
```

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+            _logAccountDetails = options?.Diagnostics?.IsAccountIdentifierLoggingEnabled ?? false;
+            _pipeline = pipeline;
+            _processService = processService ?? ProcessService.Default;
+            TenantId = options?.TenantId;
+            AdditionallyAllowedTenantIds = TenantIdResolver.ResolveAddionallyAllowedTenantIds(options?.AdditionallyAllowedTenantsCore);
+            AzdCliProcessTimeout = options?.AzdCliProcessTimeout ?? TimeSpan.FromSeconds(13);
+        }
+
+        /// <summary>
+        /// Obtains a access token from Azure Developer CLI credential, using this access token to authenticate. This method called by Azure SDK clients.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:75]:
```suggestion
        /// Obtains an access token from Azure Developer CLI credential, using this access token to authenticate. This method called by Azure SDK clients.
```

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+        /// <summary>
+        /// Create an instance of CliCredential class.
+        /// </summary>
+        public AzureDeveloperCliCredential()
+            : this(CredentialPipeline.GetInstance(null), default)
+        { }
+
+        /// <summary>
+        /// Create an instance of DeveloperCliCredential class.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:56]:
```suggestion
        /// Create an instance of the <see cref="AzureDeveloperCliCredential"/> class.
```

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+        private readonly CredentialPipeline _pipeline;
+        private readonly IProcessService _processService;
+        private readonly bool _logPII;
+        private readonly bool _logAccountDetails;
+        internal string TenantId { get; }
+        internal string[] AdditionallyAllowedTenantIds { get; }
+
+        /// <summary>
+        /// Create an instance of CliCredential class.
```
Comment (scottaddie) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:49]:
```suggestion
        /// Create an instance of the <see cref="AzureDeveloperCliCredential"/> class.
```

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            AzdCliProcessTimeout = options?.AzdCliProcessTimeout ?? TimeSpan.FromSeconds(13);
+        }
+
+        /// <summary>
+        /// Obtains a access token from Azure Developer CLI credential, using this access token to authenticate. This method called by Azure SDK clients.
+        /// </summary>
+        /// <param name="requestContext"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        //TODO 1
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:80]: Looks like we need to populate all of these empty doc comments in the file.

File: sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredential.cs [hunk 8]
```diff
@@ -15,6 +15,7 @@ namespace Azure.Identity
     /// Provides a default <see cref="TokenCredential"/> authentication flow for applications that will be deployed to Azure.  The following credential
     /// types if enabled will be tried, in order:
     /// <list type="bullet">
+    /// <item><description><see cref="AzureDeveloperCliCredential"/></description></item>
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredential.cs:18]: Is it correct that this comes first in the credential list now?

File: sdk/identity/Azure.Identity/src/DefaultAzureCredentialFactory.cs [hunk 9]
```diff
@@ -133,6 +138,22 @@ public virtual TokenCredential CreateInteractiveBrowserCredential()
...
+        {
+            var options = new AzureDeveloperCliCredentialOptions
+            {
+                TenantId = Options.TenantId,
+                AzdCliProcessTimeout = Options.DeveloperCredentialTimeout
+            };
+
+            foreach (var addlTenant in Options.AdditionallyAllowedTenants)
+            {
+                options.AdditionallyAllowedTenants.Add(addlTenant);
```
Comment (christothes) [inline sdk/identity/Azure.Identity/src/DefaultAzureCredentialFactory.cs:151]: nit:
```suggestion
            foreach (var additionalTenant in Options.AdditionallyAllowedTenants)
            {
                options.AdditionallyAllowedTenants.Add(additionalTenant);
```

File: sdk/identity/Azure.Identity/tests/CredentialTestHelpers.cs [hunk 10]
```diff
@@ -37,6 +37,27 @@ public static (string Token, DateTimeOffset ExpiresOn, string Json) CreateTokenF
...
+        }
+
+        public static (string Token, DateTimeOffset ExpiresOn, string Json) CreateTokenForAzureDeveloperCliExpiresIn(int seconds = 30)
+        {
+            var expiresOn = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(seconds);
+            var token = TokenGenerator.GenerateToken(Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), Guid.NewGuid().ToString(), expiresOn.UtcDateTime);
+            var json = $"{{ \"token\": \"{token}\", \"expiresIn\": {seconds} }}";
+            return (token, expiresOn, json);
+        }
+
```
Comment (christothes) [inline sdk/identity/Azure.Identity/tests/CredentialTestHelpers.cs:60]: If this is the same as the above CreateTokenForAzureCli* methods, let's just rename them to something more generic and re-use.

File: sdk/storage/Azure.Storage.Blobs.Batch/api/Azure.Storage.Blobs.Batch.net6.0.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
+namespace Azure.Storage.Blobs.Specialized
```
Comment (christothes) [inline sdk/storage/Azure.Storage.Blobs.Batch/api/Azure.Storage.Blobs.Batch.net6.0.cs:1]: Did you mean to include these storage changes?

File: sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredential.cs (hunk 8 above)
Comment (scottaddie) [reply sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredential.cs:18]: I was surprised by that too. I think `EnvironmentCredential` should still be first in line.

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+                bool isWinError = exception.Message.StartsWith(WinAzdCliError, StringComparison.CurrentCultureIgnoreCase);
+
+                bool isOtherOsError = AzdNotFoundPattern.IsMatch(exception.Message);
+
+                if (isWinError || isOtherOsError)
+                {
+                    throw new CredentialUnavailableException(AzdCliNotInstalled);
+                }
+
+                bool isAADSTSError = exception.Message.Contains("AADSTS");
```
Comment (KrzysztofCwalina) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:140]: Can we store this information in the exception at the point where we throw it? I worry that such string matches won't be 100% reliable.

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+            string tenantId = TenantIdResolver.Resolve(TenantId, context, AdditionallyAllowedTenantIds);
+
+            GetFileNameAndArguments(context.Scopes, tenantId, out string fileName, out string argument);
+            ProcessStartInfo processStartInfo = GetAzureDeveloperCliProcessStartInfo(fileName, argument);
+            using var processRunner = new ProcessRunner(_processService.Create(processStartInfo), AzdCliProcessTimeout, _logPII, cancellationToken);
+
+            string output;
+            try
+            {
+                output = async ? await processRunner.RunAsync().ConfigureAwait(false) : processRunner.Run();
```
Comment (KrzysztofCwalina) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:123]: What if this code executes on the server? Can it?

**Comment by KrzysztofCwalina** [comment]
I am concerned with credentials that spawn processes. I think it would be good to have an architecture meeting discussion about this design.

**Review by KrzysztofCwalina — CHANGES_REQUESTED** [review]
I would love to discuss in person why we need to be spawning processes.

**Comment by christothes** [comment]
> I would love to discuss in person why we need to be spawning processes.

Azure CLI, PowerShell, and Visual Studio credentials also do this.

**Comment by KrzysztofCwalina** [comment]
> Azure CLI, PowerShell, and Visual Studio credentials also do this.

These are apps we don't own. For azd, why can't we have a library that we can call in-proc?

**Comment by schaabs** [comment]
> These are apps we don't own. For azd, why can't we have a library that we can call in-proc?

Azure Developer CLI is written in Go. While it seems possible to compile this down to a native .dll or .so file there would be drawbacks. First we would need to package the native dependency, rather than relying on the executable, which already must have been installed, being in PATH. Also, we would need to limit the library to c-types, and deal with the native interop in all 4 languages this is being implemented in. The developer experience would be degraded in the same way it is for any native dependency. Given that we already use subprocess execution for all the other development time credentials, I think it makes the most sense here.

File: sdk/storage/Azure.Storage.Blobs.Batch/api/Azure.Storage.Blobs.Batch.net6.0.cs (hunk 11 above)
Comment (christothes) [reply sdk/storage/Azure.Storage.Blobs.Batch/api/Azure.Storage.Blobs.Batch.net6.0.cs:1]: I don't think these changes should be related to the credential changes. Perhaps your local branch is out of sync with main?

**Review by KrzysztofCwalina — APPROVED** [review]

**Comment by KrzysztofCwalina** [comment]
I had an offline discussion with @schaabs. We agreed to ship a preview as-is and then dig into pros and cons of the alternative approaches later.

File: sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    /// Enables authentication to Azure Active Directory using Azure Developer CLI to obtain an access token.
+    /// </summary>
+    public class AzureDeveloperCliCredential : TokenCredential
+    {
+        internal const string AzdCliNotInstalled = "Azure Developer CLI could not be found.";
+        internal const string AzdNotLogIn = "Please run 'azd login' from a command prompt to authenticate before using this credential.";
+        internal const string WinAzdCliError = "'azd' is not recognized";
+        internal const string AzdCliTimeoutError = "Azure Developer CLI authentication timed out.";
+        internal const string AzdCliFailedError = "Azure Developer CLI authentication failed due to an unknown error.";
+        internal const string Troubleshoot = "Please visit https://aka.ms/azure-dev for installation instructions and then, once installed, authenticate to your Azure account using 'azd login'.";
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/src/Credentials/AzureDeveloperCliCredential.cs:30]: Not a blocker for preview, but we should update TROUBLESHOOTING.md with a section for this credential, and create an  aka.ms link to it to include in the error message.  Added issue https://github.com/Azure/azure-sdk-for-net/issues/33693 to track.

**Review by schaabs — APPROVED** [review]

**Pull request merged**